feat(learner+class_category) : Add recommended route for learner and added class_category support for learners and classes

ParkornT opened the pull request:
Model:
-Learner : added Interested field which has many to many relationship between learners and class categories 
-Class_category : added support for many-to-many relationships between class and learners categories (same as Learner) 
-Class : added many-to-many relationship between Class and Class_sessions

Handlers:
-learner_handler : 1) modified GetLearners/GetLearners to expand class_sessions field
             Added routes - GET /learners/:id/recommended : return Enrollable classes that matches/not matches Learner's interests
                                    - POST /learners/:id/interests : Add class category IDs to learner's interests
                                    - DELETE /learners/:id/interests : Remove class category IDs from learner's interests
                                    - GET /learners/:id/interests : return Learner's interests
-class_handler : GET /classes : modified for Enrollable(Open) filter support, Detailed filter support
             Added routes : - GET /classes/:id/categories : return class categories with class ID
                                      - POST /classes/:id/categories : Add class category IDs to class
                                      - DELETE /classes/:id/categories : Remove class categories IDs from class
closes #84

ParkornT committed:
feat(learner+class_category) : Add recommended route for learner and add class_category support for learners and classes

Model:
-Learner : added Interested field which has many to many relationship between learners and class categories
-Class_category : added support many-to-many relationships between class and learners categories (same as Learner)
-Class : added many-to-many relationship between Class and Class_sessions

Handlers:
-learner_handler : 1) modified GetLearners/GetLearners to expand class_sessions field
             Added routes - GET /learners/:id/recommended : return Enrollable classes that matches/not matches Learner's interests
                                   - POST /learners/:id/interests : Add class category IDs to learner's interests
                                   - DELETE /learners/:id/interests : Remove class category IDs from learner's interests
                                   - GET /learners/:id/interests : return Learner's interests
-class_handler : GET /classes : modified for Enrollable(Open) filter support, Detailed filter support
                          Added routes : GET /classes/:id/categories : return class categories with class ID
                                                   POST /classes/:id/categories : Add class category IDs to class
                                                   DELETE /classes/:id/categories : Remove class categories IDs from class

ParkornT committed:
fix(handlers) : Clear many-to-many associations on delete handlers

Added logic to clear many-to-many associations when deleting Class_Category, Class, and Learner in handler

ParkornT committed:
fix: typo in comment in DeleteClass handler

Corrected comment from 'wuth' to 'with' becuase spell checker can't stand it

ParkornT committed:
fix(handlers_tests) : modified Handler Tests for new many-to-many relationship and added test cases for new routes

Fix  - user_handler : DeleteUser : now handles many-to-many relationship from Learner
       - test_handler : Fix class_handler_test, learner_handler_test, user_handler_test and class_category_handler_test to account for the new many-to-many relationship(GET/DELETE routes)
Added : - learner_handler_test : Introduces tests for adding, deleting, fetching from learner interests
              - class handler_test :  Introduces tests for adding, deleting, fetching from class_categories in classes
              - Test_Helper : Added 2 SQL mock helpers for association queries

ParkornT committed:
fix(learner_handler_test) : Update learner tests

Replaces ExpSelectByIDFound with explicit sqlmock.ExpectQuery for the full SELECT statement on learners, including deleted_at filter and ordering. This improves test accuracy by matching the actual query generated by GORM.

ParkornT committed:
fix(test) : Update deletion tests for the older tests and commented newer tests with errors

-Update deletion tests for TestDeleteClassCategory, TestDeleteLearner and added TX helpers
-removed new learner interests test fields for now

NareMu committed:
Revert "fix(test) : Update deletion tests for the older tests and commented newer tests with errors"

This reverts commit 9f439cc2802f56fcc6fc318446a40e678c1d5153.

NareMu committed:
Revert "fix(learner_handler_test) : Update learner tests"

This reverts commit 67df99f439c1fa6bafe70b51027da5ba648d8dc9.

NareMu committed:
Revert "fix(handlers_tests) : modified Handler Tests for new many-to-many relationship and added test cases for new routes"

This reverts commit 745a97b9b568470718717175a84e13df2b0fc4c0.

NareMu commented on handlers/user_handler.go:
```diff
@@ -269,6 +269,15 @@ func DeleteUser(c *fiber.Ctx) error {
 		return c.Status(500).JSON(err.Error())
 	}
+	// clear Learner's many-to-many interests
+	var lr models.Learner
```
This section is redundant because below section is already do this due to user already select all relate associations and it cascade delete the user table and relate association
```
if err := db.Select(clause.Associations).Delete(&user).Error; err != nil {
		return c.Status(500).JSON(err.Error())
```

NareMu commented on handlers/test_helper_test.go:
```diff
@@ -203,6 +227,330 @@ func ExpInsertReturningID(table string, id uint64) Exp {
+			WithArgs(args...).
+			WillReturnRows(sqlmock.NewRows([]string{"learner_id", "class_category_id"}))
+	}
+}
```
Exppreload already exist  with more generalize functions

NareMu commented on handlers/user_handler.go:
```diff
@@ -269,6 +269,15 @@ func DeleteUser(c *fiber.Ctx) error {
+		}
+	} else if !errors.Is(err, gorm.ErrRecordNotFound) {
+		return c.Status(500).JSON(err.Error())
+	}
```
this section is redundant due to below code has already do this 
```
if err := db.Select(clause.Associations).Delete(&user).Error; err != nil {
		return c.Status(500).JSON(err.Error())
```

NareMu committed:
refactor(user_handler): in DeleteUser,add new preload 'learner.interest' for cascade delete

NareMu committed:
test(helper): add new helper to Expect query about association and join table

NareMu committed:
test(helper): add new helper to Expect query about user cascade delete

NareMu committed:
test(handler): apply new expect query to test new api feature

NareMu committed:
test(handler): add new integration test to test new api feature

ParkornT committed:
Merge pull request #5 from NareMu/test/test-for-recommeded-feature

Test/test for recommended feature

ParkornT committed:
Fix(learner) : fix struct field alignment in integration test

Removed unnecessary spacing in struct field definitions for noInterestResp and noMatchResp in learner_handler_integration_test.go

ParkornT committed:
chores(classes + learner) : Added Swagger documentation

Fixes:
	-GET /classes : No longer display deleted classes or deleted fields

Added Models for Swagger:
	-ClassCategoryIdsDoc : Array of class category IDs
	-ClassCategoriesDoc : Array of class categories
	-RecommendedClassesDoc : 3 fields : RecommendedFound, RecommendedClasses, RemainingClasses

Added Swagger Documents for routes:
        -GET /classes : Now has updated fields for filter parameters
	-GET /learner/:id/recommended
	-GET,POST,DELETE /learner/:id/interests
	-GET,POST,DELETE /classes/:id/categories

ParkornT committed:
fix(class) : fix merge conflict

Introduces a 'sort' query parameter to the class listing API, allowing sorting by popular(was added by Pomtua), rating, newest, or alphabetical order

ParkornT committed:
Merge branch 'main' into feat/learner

ParkornT committed:
fix(swagger) : Update swagger.yaml

krtchnt reviewed (approved):
Thank you for the PR. :)
